design-proposals/golden-image-delivery-and-update-pipeline.md

```diff
@@ -0,0 +1,272 @@
+
+To close this usability gap between Iaas platforms and KubeVirt, the KubeVirt ecosystem should provide the tools necessary for supporting a similar pattern of disk image availability within the cluster.
+
+## Goals
+
+* Pipeline capable of importing and keeping golden images up-to-date
+* Zero operational downtime during disk updates
+* No mutating existing PVCs during the import/update flow
+* Compatible with DataVolumes
+* Compatible with Templates (meaning helm and ocp templates)
+* Transitions seamlessly with PVC Volume Populators in the future
+
+## Non Goals
+
+* Not a pipeline for updating disks attached to existing VMs
+
+## User Stories
+
+* As a user creating VMs, I want commonly used operating system disk images to automatically be made available to me without requiring uploading and managing my own os boot disk images.
+* As a user creating VMs from a template, I want to be able to stamp out a new VM from a template which will always use the latest operating system version available.
+* As a cluster admin managing VM images, I want the ability to automate sourcing in operating system disk image updates and make them available to users.
+
+## Repos
+
+* **CDI**: Provides API (tentatively called a VolumeReference) to act as a mutable symbolic link to a PVC and adds VolumeReferences as a DataVolume source type. 
+* **CDI**: Provides API (tentatively called VolumeImportPoller) to handle recurring polling/importing disk images as PVCs into a golden image namespace using the CDI import flows
+* **SSP**: Manages setting up VolumeImportPollers objects with containerDisk image sources for commonly supported operating systems.
+* **Common Templates**: Transitions to use CDI’s VolumeReferences as DataVolumeTemplate source which ensures templates always reference the most up-to-date operating system disk images.
+* **HCO**: Pass in golden image container registries to ssp operator in order to setup automation import pipeline.
+* **OCP UI**: Needs to ensure VolumeReference is linked properly to a PVC in order to determine whether boot source for template is available. 
+
+
+# Design
+
+![Golden Image Importer](golden-image-delivery-images/importer-design.png "Golden Image Importer Design")
+
+A **VolumeImportPoller** object results in its corresponding controller polling for updates to a container image on a recurring interval (every so many hours). When updates to a container image are detected, the VolumeImportPoller will automatically import the contents into a new PVC (using the standard CDI import flows) residing in a specific namespace with a unique name. The VolumeImportPoller may need to implement a PVC garbage collection policy for removing outdated PVCs after an expiration period has been exceeded.
```

Comment:
If this is only for polling/populating from registries I wonder if `registry` should be in the name?  And since this is something that happens on a schedule, I wonder if `cron` or `job` should be in the name.

Reply:
> If this is only for polling/populating from registries

just like VolumeReference, i'm unsure if i want to commit all in on this being a containerDisk thing. I'd like to leave it open for other import sources as well. I could see it making sense for http endpoints possibly in the future, which is why i structured the API for the poller to have a "source" field.

> And since this is something that happens on a schedule, I wonder if cron or job should be in the name.


I'm not opposed `VolumeImportCronJob`. There's already a precedence for the term `CronJob` in k8s, https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/, so that's probably more accurate. I think i'd want to change the api to replace `pollInterval:` with a `schedule: ` field similar to what's in the CronJob object if we do this.

What do you think about changing the naming to `VolumeImportCronJob`?

Reply:
Michael and I talked about this. We're going to use the term `DataImportCron` in order to align with the new `DataSource` object

Reply:
I'm good with the `Cron` name as long as we also support an interval syntax matching what people are used to for a cron.
`ImportPoller` seemed pretty accurate to me as that is exactly what the resource does, it polls the source for changes and imports them. That it does that regularly on a schedule is one way I could think about. I can also think of a /trigger endpoint in the future. I don't mind a different prefix.